File: Day5CSharp/Day5.cs
namespace Day5CSharp;

public class Day5
{
    public static int CalculateVentOverlapValue(IEnumerable<string> rawCoordPairs)
    {
        var ventCoordsDict = new Dictionary<Coord, int>();
        var parsedCoordPairs = new List<List<Coord>>();
        foreach (string rawCoordPair in rawCoordPairs)
        {
            string[] coordPoints = rawCoordPair.Split(" -> ");
            parsedCoordPairs.Add(new List<Coord>{Coord.CreateCoord(coordPoints.First()), Coord.CreateCoord(coordPoints.Last())});
        }

        foreach (List<Coord> parsedCoordPair in parsedCoordPairs)
        {
            IEnumerable<Coord> lineCoords = Coord.GetNonDiagCoordsLine(parsedCoordPair.First(), parsedCoordPair.Last());
            foreach (Coord lineCoord in lineCoords)
            {
                if (ventCoordsDict.ContainsKey(lineCoord))
                {
                    ventCoordsDict[lineCoord]++;
                }
                else
                {
                    ventCoordsDict.Add(lineCoord, 1);
                }
            }
        }

        return ventCoordsDict.Count(c => c.Value > 1);
    }

    public struct Coord
    {
        public int X { get; private set; }
        public int Y { get; private set; }

        public static Coord CreateCoord(string rawCoord)
        {
            var returnCoord = new Coord();
            
            string[] coordElements = rawCoord.Split(',');
            returnCoord.X = int.Parse(coordElements[0]);
            returnCoord.Y = int.Parse(coordElements[1]);

            return returnCoord;
        }

        public static Coord CreateCoord(int x, int y)
        {
            var returnCoord = new Coord();
            
            returnCoord.X = x;
            returnCoord.Y = y;

            return returnCoord;
        }

        public static IEnumerable<Coord> GetNonDiagCoordsLine(Coord coord1, Coord coord2)
        {
            var line = new List<Coord>();
            
            if (coord1.X != coord2.X && coord1.Y != coord2.Y)
            {
                return line;
            }

            if (coord1.X < coord2.X)
            {
                line = Enumerable.Range(coord1.X, coord2.X - coord1.X + 1).Select(x => Coord.CreateCoord(x, coord1.Y)).ToList();
            }
            else if (coord2.X < coord1.X)
            {
                line = Enumerable.Range(coord2.X, coord1.X - coord2.X + 1).Select(x => Coord.CreateCoord(x, coord1.Y)).ToList();
            }
            else if (coord1.Y < coord2.Y)
            {
                line = Enumerable.Range(coord1.Y, coord2.Y - coord1.Y + 1).Select(y => Coord.CreateCoord(coord1.X, y)).ToList();
            }
            else if (coord2.Y < coord1.Y)
            {
                line = Enumerable.Range(coord2.Y, coord1.Y - coord2.Y + 1).Select(y => Coord.CreateCoord(coord1.X, y)).ToList();
            }
            
            return line;
        }
    }
}

File: Day5/Day5.vb
Public Class Day5
    Public Shared Function CalculateVentOverlapValue(rawCoordPairs as IEnumerable(Of String)) As Int32
        Dim ventCoordsDict = new Dictionary(Of Coord, Integer)()
        For Each rawCoordPair As String In rawCoordPairs
            Dim coordPoints = rawCoordPair.Split(" -> ")
            Dim lineCoords = Coord.GetNonDiagCoordsLine(Coord.CreateCoord(coordPoints.First()), Coord.CreateCoord(coordPoints.Last()))
            For Each lineCoord As Coord in lineCoords
                If(ventCoordsDict.ContainsKey(lineCoord)) Then
                    ventCoordsDict.Item(lineCoord) += 1
                Else
                    ventCoordsDict.Add(lineCoord, 1)
                End If
            Next
        Next
        
        Return ventCoordsDict.Where(Function (c) c.Value > 1).Count
    End Function
End Class

Public Structure Coord
    Private _x As Int32
    Private _y As Int32

    public Shared Function CreateCoord(rawCoord As String) As Coord

        Dim returnCoord = New Coord()
        
        Dim coordElements As List(Of String) = rawCoord.Split(",").ToList()
        returnCoord._x = Int32.Parse(coordElements.Item(0))
        returnCoord._y = Int32.Parse(coordElements.Item(1))

        Return returnCoord
    End Function

    Private Shared Function CreateCoord(x As Integer, y As Integer) As Coord
        Dim returnCoord = new Coord()
        
        returnCoord._x = x
        returnCoord._y = y

        Return returnCoord
    End Function

    Public Shared Function GetNonDiagCoordsLine(coord1 As Coord, coord2 As Coord) As IEnumerable(Of Coord)
        Dim line = new List(Of Coord)()
        
        If(coord1._x <> coord2._x And coord1._y <> coord2._y) Then
            Return line
        End If
        
        If (coord1._x < coord2._x) Then
            line = (From x In Enumerable.Range(coord1._x, coord2._x - coord1._x + 1) Select Coord.CreateCoord(x, coord1._y)).ToList()
        Else If (coord2._x < coord1._x) Then
            line = (From x In Enumerable.Range(coord2._x, coord1._x - coord2._x + 1) Select Coord.CreateCoord(x, coord1._y)).ToList()
        Else If (coord1._y < coord2._y) Then
            line = (From y In Enumerable.Range(coord1._y, coord2._y - coord1._y + 1) Select Coord.CreateCoord(coord1._x, y)).ToList()
        Else If (coord2._y < coord1._y) Then
            line = (From y In Enumerable.Range(coord2._y, coord1._y - coord2._y + 1) Select Coord.CreateCoord(coord1._x, y)).ToList()
        End If
        
        Return line
    End Function

End Structure
